[Serve] Add experimental pipeline docs (#20292)
ray-project/ray: doc/source/serve/pipeline.rst

.. _serve-pipeline-api:

Pipeline API (Experimental)
===========================

This section should help you:

- understand the experimental pipeline API

> - build on top of the API to construct your multi-model inference pipelines.

Reviewer: inconsistent period usage

.. note::
    This API is experimental and the API is subject to change.

>     We are actively looking for feedback via the Ray Slack or Github issues.

Reviewer: add links for these
```suggestion
    We are actively looking for feedback via the Ray Slack or Github issues.
```

Basic API
---------

Serve Pipeline is a standalone package that can be used without Ray Serve. 

> However, the expected usage is to use it inside your Serve deployment.

Reviewer: kind of a weird thing to open with. I'd start with the problem statement: This package is mean to make it seamless to develop and deploy multi-model pipelines. Add a sentence or two about what the advantages are (all in code, good perf)

You can import it as:

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __import_start__
    :end-before: __import_end__

> You can decorate any function or classes using ``pipeline.step``. You should
> then pass it's upstream input by calling the decorated function or class. In

Reviewer: ```suggestion
You can decorate any function or class using ``pipeline.step``. You can then
combine these steps into a pipeline by calling the decorated steps. In
```

> the example below, we have a single step that takes the special node ``pipeline.INPUT``.

Reviewer: ```suggestion
the example below, we have a single step that takes the special node ``pipeline.INPUT``, which is a placeholder for the arguments that will be passed into the pipeline.
```

> Once you created the pipeline node, you can call ``.deploy()`` to instantiate the whole pipeline.

Reviewer: ```suggestion
Once you have defined the pipeline by combining one or more steps, you can call ``.deploy()`` to instantiate it.
```

Once you have instantiated the pipeline, you can call ``.call(inp)`` to invoke synchronously. 

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __simple_pipeline_start__
    :end-before: __simple_pipeline_end__

The input to a pipeline node can be the ``pipeline.INPUT`` special node or 
one or more other pipeline nodes. Here is an example of simple chaining pipeline. 

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __simple_chain_start__
    :end-before: __simple_chain_end__

> For classes, you need to instantiate them with init args first, then pass in their
> upstream nodes. 

Reviewer: ```suggestion
For classes, you need to instantiate them with init args first, then pass in their
upstream nodes.  This allows you to have the same code definition but pass different arguments, like URIs for model weights.
```

Reviewer: we should add an example for that

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __class_node_start__
    :end-before: __class_node_end__

The decorator also takes two arguments to configure where the node will be executed.

.. autofunction:: ray.serve.pipeline.step

> Here is an example pipeline that uses actors instead of local execution mode.

Reviewer: say that local execution mode is useful for testing?

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __pipeline_configuration_start__
    :end-before: __pipeline_configuration_end__

Chaining Example
----------------

In this section, we show how to implement a two stage pipeline that's common

> for computer vision task. For workload like image classification, the preprocessing

Reviewer: ```suggestion
for computer vision tasks. For workloads like image classification, the preprocessing
```

steps are CPU bounded and hard to parallelize. The actual inference steps can run

> on GPU and batched. They are often split up into separate stages and scaled separately

Reviewer: why do I want to batch it?

to increase throughput.

.. literalinclude:: ../../../python/ray/serve/examples/doc/snippet_pipeline.py
    :language: python
    :start-after: __preprocessing_pipeline_example_start__
    :end-before: __preprocessing_pipeline_example_end__

> Ensemble Example

Reviewer: can you add a section after this listing more potential use cases?

Also, can we add a simple example of "combining business logic + ML"? This could be just calling out that you can do anything in the "combine output" step
